Have the process survive parse errors in individual files

gurgeous opened the pull request:
How about this? #83

mislav commented on lib/ripper-tags/parser.rb:
```diff
@@ -341,6 +354,7 @@ def process(sexp)
       when Array
         sexp.each{ |child| process(child) }
       when Symbol
+        @sexp = sexp
```
Why is this assignment here?

mislav commented on test/fixtures/bad_code.rb:
```diff
@@ -0,0 +1,32 @@
+#
+# Note: This file is from devise. It was encountered in the wild and was
```
Can we distill down the fixture file to a bare minimum enough to reproduce the erb failure? Then we can get rid of the explanation comment because the code wouldn't come from devise anymore.

Also, can this file be called `erb_template.rb` instead of `bad_code.rb` so it's a little bit more descriptive?

mislav commented on test/test_data_reader.rb:
```diff
@@ -137,6 +139,31 @@ def test_input_file_as_stdin
     end
   end
 
+  def test_errors
```
`test_errors` seems large and that tells me that it's testing too many different scenarios and that it could probably broken down into multiple test cases

mislav commented on test/test_data_reader.rb:
```diff
@@ -156,4 +183,14 @@ def ignore_warnings
+    old_stderr = $stderr
+    $stderr = StringIO.new
+    begin
+      yield
```
Could `capture_stderr` return `$stderr.string` so tests can verify is something was written to stderr or not?

mislav commented on lib/ripper-tags/data_reader.rb:
```diff
@@ -130,20 +132,33 @@ def normalize_encoding(str)
+
+          # print sexp and unwrap the RuntimeError
+          if err.is_a?(Visitor::Error)
+            sexp = err.sexp.map { |i| i.is_a?(Array) ? "[...]" : i }
```
I don't think it's necessary to print the sexp. Nobody will understand it unless they're literally working on ripper-tags code. I think that would alleviate the need for having Visitor::Error.

It would be great if we could print the line of the file on which the error has ocurred, but that won't be possible without significantly refactoring Parser and Visitor to always preserve line numbers, no matter what happens.

gurgeous commented:
Sure, I'm always open to feedback.

- removed custom error / sexp printing
- renamed erb fixture, shrank it. I attempted this before but tried much hard this time. :)
- turned one test into three
- made capture_stderr match capture_stdout

gurgeous commented:
I think this will do the trick.

gurgeous commented:
Third time is the charm... Comments:

- Modified RUBY_VERSION check
- I don't think we can make those tests do anything useful if RUBY < 2.3. I can't make that template crash with those versions of Ripper. Maybe we can address this in another PR if we find a different crash?
- Added stderr asserts
- Fixed comment in erb_template.rb
- The `self.down` is required - the crash depends on it

mislav commented on lib/ripper-tags/data_reader.rb:
```diff
@@ -130,20 +132,25 @@ def normalize_encoding(str)
+          @error_count += 1
+
+          # Die if we ran without --force and we haven't seen any tags yet.
+          if !options.force && @tag_count == 0
```
One final thought about this logic: suppose you run `ripper-tags on a `lib/` directory which has 2 files in it: `good.rb` and `erb_template.rb`. The first one will generate some tags, but the second one will enounter a parse error. Because there are already some tags, this condition will *not* halt processing, but just print to stderr and eventually return with exit status 2.

However, let's say that `erb_template.rb` gets processed before `good.rb`. Now the processing halts immediately because the first file had a parse error before any tags were able to get generated. The behavior of ripper-tags is suddenly very different depending on the order of input files.

I know that checking the number of tags is something that I've suggested initially, but now I don't think it's such a great idea. Let's pick consistency and simplicity. How about forgiving all parse errors (while outputting them to stderr) regardless of the number of tags being collected at the time of encountering the error? Let me know your thoughts.

gurgeous commented:
Sure. One question - do we keep --force? After this change the only thing
it will do is change the exit code from 2 to 0 when errors are encountered.

If we decide to kill that functionality we probably need to leave it in for
backward compatibility.

On Thu, Mar 29, 2018 at 7:24 AM, Mislav Marohnić <notifications@github.com>
wrote:

> *@mislav* commented on this pull request.
> ------------------------------
>
> In lib/ripper-tags/data_reader.rb
> <https://github.com/tmm1/ripper-tags/pull/85#discussion_r178072279>:
>
> >        file_finder.each_file do |file|
>          begin
>            $stderr.puts "Parsing file #{file}" if options.verbose
>            extractor = tag_extractor(file)
>          rescue => err
> -          if options.force
> -            $stderr.puts "Error parsing `#{file}': #{err.message}"
> -          else
> +          # print the error
> +          $stderr.puts "Error parsing `#{file}': #{err.message}"
> +          @error_count += 1
> +
> +          # Die if we ran without --force and we haven't seen any tags yet.
> +          if !options.force && @tag_count == 0
>
> One final thought about this logic: suppose you run ripper-tags on alib/directory
> which has 2 files in it:good.rbanderb_template.rb`. The first one will
> generate some tags, but the second one will enounter a parse error. Because
> there are already some tags, this condition will *not* halt processing,
> but just print to stderr and eventually return with exit status 2.
>
> However, let's say that erb_template.rb gets processed before good.rb.
> Now the processing halts immediately because the first file had a parse
> error before any tags were able to get generated. The behavior of
> ripper-tags is suddenly very different depending on the order of input
> files.
>
> I know that checking the number of tags is something that I've suggested
> initially, but now I don't think it's such a great idea. Let's pick
> consistency and simplicity. How about forgiving all parse errors (while
> outputting them to stderr) regardless of the number of tags being collected
> at the time of encountering the error? Let me know your thoughts.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tmm1/ripper-tags/pull/85#pullrequestreview-108068124>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgHc0GpM_7ZUtxL2b8xf--SSjJPi2GLks5tjO6QgaJpZM4S56wc>
> .
>

mislav commented:
We keep `--force` for backwards compatibility but change its documentation re: exit code.

gurgeous commented:
How about this? The only additional change I might suggest would be to update the README to document exit code 2. We can do that later, though. That file probably needs a refresh anyway.

mislav commented:
I've pushed some changes. I've opted against exit status 2 for now. The new logic: exit with status 1 if there were as many parse errors as files (i.e. if all files have failed). Otherwise, exit with 0.

If some tool author requests to be able to distinguish between these cases, we can update exit codes. But for now, I vote to keep it simple.

I've also renamed your PR title to something more descriptive and moved the issue reference from the title (where it's invalid and not a real link) to the PR body so it properly links to the corresponding issue. This is something that I suggest you keep in mind for future PRs. 😉 

Thank you for all your help and patience! Would you kindly give your review to my updates? 🙇

skarger commented:
@gurgeous @mislav Just wanted to chime in to keep this PR from slipping through the cracks, since I'd benefit from the feature being released. Thanks for your work.

gurgeous commented:
Oh, sorry about that! I missed that I was supposed to review. Thanks for the nudge. Give me a sec here...

mislav commented:
@skarger This is now v0.6.0.
